pkg/clusterinfo/clusterinfo.go

```diff
@@ -0,0 +1,276 @@
+	if _, registered := r.consumerChannels[c.String()]; registered {
+		return ErrConsumerAlreadyRegistered
+	}
+	r.consumerChannels[c.String()] = c.ClusterLabelUpdates()
+	return nil
+}
+
+// Run starts the update loop and periodically queries the / endpoint
+// The update loop is terminated upon ctx cancellation. The call blocks until the first
+// call to the cluster info endpoint was successful
+func (r *Retriever) Run(ctx context.Context) error {
+	startupComplete := make(chan struct{})
+	// start update routine
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				_ = level.Info(r.logger).Log(
+					"msg", "context cancelled, exiting cluster info update loop",
+					"err", ctx.Err(),
+				)
+				return
+			case <-r.sync:
+				_ = level.Info(r.logger).Log(
+					"msg", "providing consumers with updated cluster info label",
+				)
+				res, err := r.fetchAndDecodeClusterInfo()
+				if err != nil {
+					_ = level.Error(r.logger).Log(
+						"msg", "failed to retrieve cluster info from ES",
+						"err", err,
+					)
+					r.updateMetrics(nil)
+					continue
+				}
+				r.updateMetrics(res)
+				for name, consumerCh := range r.consumerChannels {
+					_ = level.Debug(r.logger).Log(
+						"msg", "sending update",
+						"consumer", name,
+						"res", fmt.Sprintf("%+v", res),
+					)
+					*consumerCh <- res
+				}
+				// close startupComplete if not already closed
+				select {
+				case <-startupComplete:
+				default:
+					close(startupComplete)
+				}
+			}
+		}
+	}(ctx)
+	// trigger initial cluster info call
+	_ = level.Info(r.logger).Log(
+		"msg", "triggering initial cluster info call",
+	)
+	r.sync <- struct{}{}
+
+	// start a ticker routine
+	go func(ctx context.Context) {
+		if r.interval <= 0 {
+			_ = level.Info(r.logger).Log(
+				"msg", "no periodic cluster info label update requested",
+			)
+			return
+		}
+		ticker := time.NewTicker(r.interval)
+		for {
+			select {
+			case <-ctx.Done():
+				_ = level.Info(r.logger).Log(
+					"msg", "context cancelled, exiting cluster info trigger loop",
+					"err", ctx.Err(),
+				)
+				return
+			case <-ticker.C:
+				_ = level.Debug(r.logger).Log(
+					"msg", "triggering periodic update",
+				)
+				r.sync <- struct{}{}
+			}
+		}
+	}(ctx)
+
+	// block until the first retrieval was successful
+	select {
+	case <-startupComplete:
+		// first sync has been successful
+		level.Debug(r.logger).Log("msg", "initial clusterinfo sync succeeded")
+		return nil
+	case <-time.After(initialTimeout):
+		// initial call timed out
+		return ErrInitialCallTimeout
+	case <-ctx.Done():
+		// context cancelled
+		return nil
+	}
+}
+
+func (r *Retriever) fetchAndDecodeClusterInfo() (*Response, error) {
+	var response *Response
+	u := *r.url
+	u.Path = path.Join(r.url.Path, "/")
+
+	res, err := r.client.Get(u.String())
+	if err != nil {
+		_ = level.Error(r.logger).Log(
+			"msg", "failed to get cluster info",
+			"err", err,
+		)
+		return nil, err
+	}
+
+	defer func() {
```

Comment:
I think it's ok to simply do `defer res.Body.Close()`

Reply:
yes, I totally agree. This clumsy construction is a victim of the newly introduced errcheck, which I consider as useful most of the time. I already had quite some "sh*t, errcheck and logging don't like each other", so I'm consider ditching errcheck again. Thoughts are welcome...

Reply:
and ... thanks for the review!

Reply:
@zwopir what is about using Prometheus common/log library? I think, this would handle your logging problem better.

Reply:
it's beyond the scope of this PR, but maybe it's worth considering. I personally don't really like logrus (which prometheus logging is based on), which of cause shouldn't prevent us changing the logging if needed.

Reply:
Prometheus itself uses the go-kit logger these days.

Reply:
@zwopir Take a look at this example which turns of parts of errcheck:
https://github.com/improbable-eng/thanos/blob/master/.errcheck_excludes.txt

Reply:
thanks! I will definitely add errcheck excludes for logging and probably resp.body.Close(). You're doomed anyways if those operations fail >:(